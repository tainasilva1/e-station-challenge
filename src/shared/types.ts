export interface IChartData {
  [key: string]: number | string;
}

export type DataKey = {
  key: string;
  color: string;
}

export interface IConsumptionData {
  id: string,
  reference: string,
  year: number,
  month: number,
  day: number,
  hour: number,
  consumption: number,
  agent: string,
  meter: string,
  origin: string,
  intervalInMinutes: number,
  isEstimated: boolean
}
