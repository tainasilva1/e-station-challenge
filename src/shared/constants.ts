export const MONTH_NAMES = ['Jan', 'Fev', 'Mar', 'Abr', 'Mai', 'Jun', 'Jul', 'Ago', 'Set', 'Out', 'Nov', 'Dez']

export const KEYS_DATA_BY_DAY = {
  CONSUMPTION: 'consumption',
  MAXIMUM_FLEX: 'maximum_flex',
  MINIMUM_FLEX: 'minimum_flex',
  FLAT_CONSUMPTION: 'flat_consumption'
}

export const KEYS_BY_LAST_WEEK = [{ key: 'consumo', color: '#8884d8'}]

export const KEYS_BY_YEAR = [
  { key: '2021', color: '#3c81f6' },
  { key: '2022', color: '#1ab8a6' }
]

export const KEYS_BY_DATE = [
  {
    key: KEYS_DATA_BY_DAY.CONSUMPTION,
    color: '#5e9ae5'
  },
  {
    key: KEYS_DATA_BY_DAY.MAXIMUM_FLEX,
    color: '#f97415'
  },
  {
    key: KEYS_DATA_BY_DAY.MINIMUM_FLEX,
    color: '#6466f1'
  },
  {
    key: KEYS_DATA_BY_DAY.FLAT_CONSUMPTION,
    color: '#23c45e'
  },
]


export const BY_DAY_LEGENDS = {
  [KEYS_DATA_BY_DAY.CONSUMPTION]: 'Consumo',
  [KEYS_DATA_BY_DAY.MAXIMUM_FLEX]: 'Flex Máxima',
  [KEYS_DATA_BY_DAY.MINIMUM_FLEX]: 'Flex Mínima',
  [KEYS_DATA_BY_DAY.FLAT_CONSUMPTION]: 'Consumo Flat'
}

export const CONSUMPTION_KEYS = {
  AGENT: 'agent',
  METER: 'meter',
  REFERENCE: 'reference',
  HOUR: 'hour',
  CONSUMPTION: 'consumption',
  ORIGIN: 'origin'
}

export const CONSUMPTION_LABELS = {
  [CONSUMPTION_KEYS.AGENT]: 'Agente',
  [CONSUMPTION_KEYS.METER]: 'Ponto',
  [CONSUMPTION_KEYS.REFERENCE]: 'Data',
  [CONSUMPTION_KEYS.HOUR]: 'Hora',
  [CONSUMPTION_KEYS.CONSUMPTION]: 'Consumo Ativo (MWh)',
  [CONSUMPTION_KEYS.ORIGIN]: 'Origem'
}