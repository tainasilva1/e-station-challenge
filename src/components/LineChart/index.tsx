import { BY_DAY_LEGENDS } from '@/src/shared/constants';
import { DataKey, IChartData } from '@/src/shared/types';
import { memo } from 'react';
import { LineChart as LineChartRecharts, XAxis, YAxis, CartesianGrid, Tooltip, Legend, Line, ResponsiveContainer } from 'recharts';
import ContentLoader from 'react-content-loader';

type Props = {
  data?: IChartData[];
  dataKey: DataKey[];
  unit?: string;
}

export const LineChart = ({
  data,
  dataKey,
  unit = '(MWh)',
}: Props) => {
  if (!data) return (
    <ContentLoader width='100%' height={300} viewBox="0 0 680 264">
      <rect x="20" y="5" rx="0" ry="0" width="3" height={300} />
      <rect x="20" y="260" rx="0" ry="0" width={640} height="3" />
      <path d="M 656 30.566 C 556.69 20.52 504.966 61.705 335 71.75 c -121.241 8.175 -198.621 66.358 -300 66.358 V 245 h 630 V 30.566 z" />
    </ContentLoader>
  )
  return (
    <ResponsiveContainer width="100%" height={300}>
      <LineChartRecharts
        data={data}
        margin={{
          top: 5,
          right: 20,
          left: 20,
          bottom: 5,
        }}
      >
        <CartesianGrid strokeDasharray="3 3" vertical={false} />
        <XAxis style={{ fontSize: '0.8rem' }} width={80} tick={false} tickLine={false} dataKey="name" />
        <YAxis style={{ fontSize: '0.8rem' }} unit={unit} tickLine={false} domain={['DataMin', 'DataMax']} />
        <Tooltip formatter={(value, name) => [`${value} ${unit}`, BY_DAY_LEGENDS[name]]} />
        <Legend
          align='right'
          verticalAlign='top'
          iconType='circle'
          height={36}
          formatter={(value) => BY_DAY_LEGENDS[value]}
        />
        {dataKey.map((data) => (
          <Line
            key={data.key}
            type="monotone"
            dataKey={data.key}
            stroke={data.color}
            strokeWidth={2}
            dot={false}
          />
        ))}
      </LineChartRecharts>
    </ResponsiveContainer>
  )
}
export default memo(LineChart);