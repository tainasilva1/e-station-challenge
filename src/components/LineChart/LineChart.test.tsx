import { KEYS_BY_DATE } from '@/src/shared/constants';
import { render } from '@testing-library/react'
import LineChart from '../LineChart';

const dataMock = [{"name":1,"consumption":106,"maximum_flex":110,"minimum_flex":90,"flat_consumption":100},{"name":2,"consumption":92,"maximum_flex":110,"minimum_flex":90,"flat_consumption":100},{"name":3,"consumption":98,"maximum_flex":110,"minimum_flex":90,"flat_consumption":100},{"name":4,"consumption":95,"maximum_flex":110,"minimum_flex":90,"flat_consumption":100},{"name":5,"consumption":93,"maximum_flex":110,"minimum_flex":90,"flat_consumption":100}]
test('renders LineChart correctly', () => {
  const { container } = render(<LineChart data={dataMock} dataKey={KEYS_BY_DATE} />)
  expect(container).toMatchSnapshot()
})