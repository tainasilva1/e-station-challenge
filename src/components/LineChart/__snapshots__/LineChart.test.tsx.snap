// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders LineChart correctly 1`] = `
<div>
  <div
    class="recharts-responsive-container"
    style="width: 100%; height: 300px; min-width: 0;"
  />
</div>
`;
